### ISSUE TITLE
computing IOU in Mask RCNN

### ISSUE THREAD

**Issue by Basma-hash** [matterport/Mask_RCNN#2440]
Hello everyone,
I am working on project which segment custom data using Mask RCNN and I want to compute the IOU for the output prediction.

can anyone help??

**Comment by egrelier** [comment]
The file utils.py contains the function compute_iou that does what you want.

**Comment by Basma-hash** [comment]
@egrelier yes I know that, but I wanna compute IOU alone after running the code on my custom data so what are the names of IOU function inputs do I need to enter?

**Comment by Basma-hash** [comment]
def compute_iou(box, boxes, box_area, boxes_area):
    """Calculates IoU of the given box with the array of the given boxes.
    box: 1D vector [y1, x1, y2, x2]
    boxes: [boxes_count, (y1, x1, y2, x2)]
    box_area: float. the area of 'box'
    boxes_area: array of length boxes_count.

    Note: the areas are passed in rather than calculated here for
    efficiency. Calculate once in the caller to avoid duplicate work.
    """
    # Calculate intersection areas
    y1 = np.maximum(box[0], boxes[:, 0])
    y2 = np.minimum(box[2], boxes[:, 2])
    x1 = np.maximum(box[1], boxes[:, 1])
    x2 = np.minimum(box[3], boxes[:, 3])
    intersection = np.maximum(x2 - x1, 0) * np.maximum(y2 - y1, 0)
    union = box_area + boxes_area[:] - intersection[:]
    iou = intersection / union
    return iou
in this code, what does area mean and how I calculate it?

**Comment by GiUKim** [comment]
did u solve this issue??

**Comment by mesakh123** [comment]
Maybe this one, only for one class prediction
```
def merged_mask(masks):
    """
    merge mask into one and return merged mask
    """
    n= masks.shape[2]
    
    if n!=0:        
        merged_mask = np.zeros((masks.shape[0], masks.shape[1]))
        for i in range(n):
            merged_mask+=masks[...,i]
        merged_mask=np.asarray(merged_mask,dtype=np.uint8)   
        return merged_mask
    return masks[:,:,0]
    

def compute_iou(predict_mask, gt_mask):
    """
    Computes Intersection over Union score for two binary masks.
    :param predict_mask: numpy array
    :param gt_mask: numpy array
    :type1 and type2 results are same
    :return iou score:
    """
    if predict_mask.shape[2]==0:
        return 0
    mask1 = merged_mask(predict_mask)
    mask2 = merged_mask(gt_mask)
    
    #type 1
    intersection = np.sum((mask1 + mask2) > 1)
    union = np.sum((mask1 + mask2) > 0)
    iou_score = intersection / float(union)
    #print("Iou 1 : ",iou_score)
    
    #type2
    intersection = np.logical_and(mask1, mask2)#*
    union = np.logical_or(mask1, mask2)# +
    iou_score = np.sum(intersection) / np.sum(union)
    #print("Iou 2 : ",iou_score)
    return iou_score

```

**Comment by buseyaren** [comment]
> Maybe this one, only for one class prediction
> 
> ```
> def merged_mask(masks):
>     """
>     merge mask into one and return merged mask
>     """
>     n= masks.shape[2]
>     
>     if n!=0:        
>         merged_mask = np.zeros((masks.shape[0], masks.shape[1]))
>         for i in range(n):
>             merged_mask+=masks[...,i]
>         merged_mask=np.asarray(merged_mask,dtype=np.uint8)   
>         return merged_mask
>     return masks[:,:,0]
>     
> 
> def compute_iou(predict_mask, gt_mask):
>     """
>     Computes Intersection over Union score for two binary masks.
>     :param predict_mask: numpy array
>     :param gt_mask: numpy array
>     :type1 and type2 results are same
>     :return iou score:
>     """
>     if predict_mask.shape[2]==0:
>         return 0
>     mask1 = merged_mask(predict_mask)
>     mask2 = merged_mask(gt_mask)
>     
>     #type 1
>     intersection = np.sum((mask1 + mask2) > 1)
>     union = np.sum((mask1 + mask2) > 0)
>     iou_score = intersection / float(union)
>     #print("Iou 1 : ",iou_score)
>     
>     #type2
>     intersection = np.logical_and(mask1, mask2)#*
>     union = np.logical_or(mask1, mask2)# +
>     iou_score = np.sum(intersection) / np.sum(union)
>     #print("Iou 2 : ",iou_score)
>     return iou_score
> ```

It worked for me, thank you!

**Comment by Yash-Vora** [comment]
Can anyone share exact code how to perform iou to remove ovelapping boxes.

**Comment by IremYoldas** [comment]
> Maybe this one, only for one class prediction
> 
> ```
> def merged_mask(masks):
>     """
>     merge mask into one and return merged mask
>     """
>     n= masks.shape[2]
>     
>     if n!=0:        
>         merged_mask = np.zeros((masks.shape[0], masks.shape[1]))
>         for i in range(n):
>             merged_mask+=masks[...,i]
>         merged_mask=np.asarray(merged_mask,dtype=np.uint8)   
>         return merged_mask
>     return masks[:,:,0]
>     
> 
> def compute_iou(predict_mask, gt_mask):
>     """
>     Computes Intersection over Union score for two binary masks.
>     :param predict_mask: numpy array
>     :param gt_mask: numpy array
>     :type1 and type2 results are same
>     :return iou score:
>     """
>     if predict_mask.shape[2]==0:
>         return 0
>     mask1 = merged_mask(predict_mask)
>     mask2 = merged_mask(gt_mask)
>     
>     #type 1
>     intersection = np.sum((mask1 + mask2) > 1)
>     union = np.sum((mask1 + mask2) > 0)
>     iou_score = intersection / float(union)
>     #print("Iou 1 : ",iou_score)
>     
>     #type2
>     intersection = np.logical_and(mask1, mask2)#*
>     union = np.logical_or(mask1, mask2)# +
>     iou_score = np.sum(intersection) / np.sum(union)
>     #print("Iou 2 : ",iou_score)
>     return iou_score
> ```

It's worked for me too, thank you!

**Comment by SHANKS-red-hue** [comment]
> 

In utils , you will see this function :non_max_suppression; 
`def non_max_suppression(boxes, scores, threshold):
    """Performs non-maximum suppression and returns indices of kept boxes.
    boxes: [N, (y1, x1, y2, x2)]. Notice that (y2, x2) lays outside the box.
    scores: 1-D array of box scores.
    threshold: Float. IoU threshold to use for filtering.
    """
    assert boxes.shape[0] > 0
    if boxes.dtype.kind != "f":
        boxes = boxes.astype(np.float32)

    # Compute box areas
    y1 = boxes[:, 0]
    x1 = boxes[:, 1]
    y2 = boxes[:, 2]
    x2 = boxes[:, 3]
    area = (y2 - y1) * (x2 - x1)

    # Get indicies of boxes sorted by scores (highest first)
    ixs = scores.argsort()[::-1]

    pick = []
    while len(ixs) > 0:
        # Pick top box and add its index to the list
        i = ixs[0]
        pick.append(i)
        # Compute IoU of the picked box with the rest
        iou = compute_iou(boxes[i], boxes[ixs[1:]], area[i], area[ixs[1:]])
        # Identify boxes with IoU over the threshold. This
        # returns indices into ixs[1:], so add 1 to get
        # indices into ixs.
        remove_ixs = np.where(iou > threshold)[0] + 1
        # Remove indices of the picked and overlapped boxes.
        ixs = np.delete(ixs, remove_ixs)
        ixs = np.delete(ixs, 0)
    return np.array(pick, dtype=np.int32)`

area = (y2 - y1) * (x2 - x1) is author calculated area method (no issue)
But something need know  . 
In compute_iou function  : 
`intersection = np.maximum(x2 - x1, 0) * np.maximum(y2 - y1, 0)`
eg: x2, x1, y2, y1:(2.0, 1.5, 1.0, 2.0)
        intersection = 0
if you want get intersection area of iou:
can use this :
       `intersection = np.abs(y2 - y1) * np.abs( x2 - x1)`
it will give you area of intersection
